File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {
    
    static void dfs(int[][] graph, int current, boolean[] isVisited, int n ){
        
        
        for(int nextNode=0; nextNode<n; nextNode++){
            if(!isVisited[nextNode]&&graph[current][nextNode]==1){
                isVisited[nextNode] = true; 
                dfs(graph, nextNode, isVisited, n); 
            }
        }
        
        
        
    }
    
    static final int NOT_VISITED = -1; 
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        boolean[] isVisited = new boolean[n];
        
        
        for(int node=0; node<n; node++){
            if(!isVisited[node]){
                isVisited[node] = true;
                
                dfs(computers, node, isVisited,n); 
                
                answer++; 
            }
            
            
            
            
        }
        
        //모든 노드 순회 
            //방문했는가?
                //방문할 때마다 개수를 세기 
                //시작점으로부터 dfs 돌기 
                    //dfs 내부에서 시작점으로 마킹해놓기 
            
        
        return answer;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp

#include <string>
#include <vector>
#include <unordered_set> 

using namespace std;

class UnionFind{
    private:
        vector<int> parents;
        vector<int> rank;
    
        void init(int n){
            parents.resize(n);
            rank.assign(n,0); 
            
            for(int node=0; node<n; node++){
                parents[node] = node; 
            }
        }
    
    public:
        UnionFind(int n){
            init(n); 
        }    
    
        int find(int node){
            if(parents[node]==node){
                return node; 
            }
            
            return parents[node] = find(parents[node]); 
        }
    
        void unite(int a, int b){
            int parentA = find(a);
            int parentB = find(b); 
            
            if(parentA==parentB){
                return; 
            }
            
            if(rank[parentA]<rank[parentB]){
                swap(parentA, parentB); 
            }
            
            parents[parentB] = parentA; 
            
            if(rank[parentA]==rank[parentB]){
                rank[parentA]++; 
            }
        }
};


int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    UnionFind network = UnionFind(n); 
    
    for(int com1=0; com1<n; com1++){
        for(int com2=0; com2<n; com2++){
            if(com1==com2)continue; 
            
            if(computers[com1][com2]==1){
                network.unite(com1, com2); 
            }
        }
    }
    
    unordered_set<int> networkTypes; 
    for(int com=0; com<n; com++){
        networkTypes.insert(network.find(com));
    }
    
    return networkTypes.size();
}